### PR TITLE
feat(ui): Improve section movement indication and add character-level diff highlighting (fixes #64, #74)

### LINKED ISSUE #64: improve section movement indication in the ui

### LINKED ISSUE #74: feat(ui): Add character-level diff highlighting in split view
## Description

Implement character-level diff highlighting in the Split View to show exactly which characters changed within a line, similar to how GitHub displays inline diffs.

### Current State
- The Split View shows old and new YAML content side-by-side
- Changed lines are highlighted with background colors (yellow/green/red)
- The Cards View shows old and new values for each change
- **Missing**: Character-level highlighting within changed lines

### Problem
When a value changes slightly (e.g., `nginx:1.19` → `nginx:1.21`), users must visually compare the old and new values to spot the difference. This is especially difficult for:
- Long strings with minor changes
- Similar numbers (e.g., `5432` vs `5433`)
- Complex nested values
- Paths or URLs with single character differences

### User Experience Goal
When viewing a change:
- The specific characters that changed should be highlighted
- Additions should be shown in green/underline
- Deletions should be shown in red/strikethrough
- Unchanged parts of the line should remain neutral
- Should work in both Split View and Cards View

### Character-Level Diff Algorithm
- [ ] Implement character-level diff algorithm (e.g., using `diff-match-patch` or similar)
- [ ] Handle string values correctly
- [ ] Handle numeric values (show digit changes)
- [ ] Handle multi-word changes efficiently

### Split View Integration
- [ ] Character-level highlighting in changed lines
- [ ] Deleted characters shown with red background/strikethrough
- [ ] Added characters shown with green background/underline
- [ ] Unchanged characters shown normally
- [ ] Works for both sides (old and new)

### Cards View Integration
- [ ] Character-level highlighting in "Old Value" and "New Value" sections
- [ ] Consistent styling with Split View
- [ ] Handle complex values (JSON/YAML formatted)

### Performance
- [ ] Efficient for large diffs (don't compute char diff for unchanged lines)
- [ ] Lazy computation (only compute when visible)
- [ ] No UI lag when scrolling through many changes

### Recommended Library
Consider using [`diff-match-patch`](https://github.com/google/diff-match-patch) (available for JS):
```bash
npm install diff-match-patch
```

Or React-specific libraries:
- [`react-diff-viewer`](https://github.com/praneshr/react-diff-viewer)
- [`diff`](https://github.com/kpdecker/jsdiff) (lightweight)

### Example Implementation

```tsx
import { diffChars } from 'diff';

function highlightCharDiff(oldStr: string, newStr: string) {
  const diff = diffChars(oldStr, newStr);
  
  return diff.map((part, index) => {
    const className = part.added 
      ? 'bg-green-200 text-green-800' 
      : part.removed 
        ? 'bg-red-200 text-red-800 line-through' 
        : '';
    return (
      <span key={index} className={className}>
        {part.value}
      </span>
    );
  });
}
```

### Files to Modify

**New Files**:
- `ui/lib/charDiff.ts` - Character diff utility functions
- `ui/components/CharDiffHighlight.tsx` - Reusable component for char-level highlighting

**Modified Files**:
- `ui/components/SplitDiffView.tsx` - Integrate char highlighting in line display
- `ui/components/ChangeCard.tsx` - Integrate char highlighting in value display
- `ui/package.json` - Add diff library dependency

### Visual Design

**Example (before)**:
```
Old: image: nginx:1.19
New: image: nginx:1.21
```

**Example (after)**:
```
Old: image: nginx:1.[19]  ← "19" has red strikethrough
New: image: nginx:1.[21]  ← "21" has green highlight
```

**CSS Classes** (Tailwind):
```css
.char-added { @apply bg-green-200 text-green-800 rounded-sm; }
.char-removed { @apply bg-red-200 text-red-800 line-through rounded-sm; }
```

### Edge Cases to Handle
1. **Whitespace changes**: Show spaces/tabs visually (e.g., `·` for space)
2. **Empty strings**: Handle `""` → `"value"` gracefully
3. **Type changes**: String to number, etc.
4. **Long values**: Truncate with ellipsis but allow expansion
5. **Multi-line values**: Handle `|` block scalars
6. **Special characters**: HTML escape or handle quotes

### Unit Tests
- [ ] Test `diffChars` utility with various inputs
- [ ] Test empty string handling
- [ ] Test whitespace handling
- [ ] Test special character handling
- [ ] Test performance with long strings

### Component Tests
- [ ] Test `CharDiffHighlight` component renders correctly
- [ ] Test split view integration
- [ ] Test cards view integration

### Visual Tests (Manual)
- [ ] Verify highlighting is visually clear
- [ ] Verify colors are accessible (color blindness)
- [ ] Verify works in dark mode (if supported)

## Documentation Updates

- [ ] Update UI component documentation
- [ ] Add example screenshots to README
- [ ] Document the char diff algorithm in developer docs

## Related Issues

- Related: #73 - Line number mapping and highlighting
- Related: #71 - Text/YAML formatters (similar highlighting concept)
- Blocks: None
- Blocked by: None (can start independently)

## Estimated Effort

**Medium (M)** - 2-3 days

| Task | Effort |
|------|--------|
| Research and select diff library | 0.5 day |
| Implement char diff utility | 0.5 day |
| Integrate into Split View | 0.5 day |
| Integrate into Cards View | 0.5 day |
| Styling and polish | 0.5 day |
| Tests and documentation | 0.5 day |

### Split View with Character Highlighting
```
┌─────────────────────────────┬─────────────────────────────┐
│ Old Version                 │ New Version                 │
├─────────────────────────────┼─────────────────────────────┤
│  image: nginx:1.⟦19⟧       │  image: nginx:1.⟦21⟧        │
│         ▲▲ red strikethrough│         ▲▲ green highlight │
└─────────────────────────────┴─────────────────────────────┘
```

### Cards View with Character Highlighting
```
┌─────────────────────────────────────────────────────────────┐
│ ~ Value Changed    spec.containers[0].image                 │
├─────────────────────────────────────────────────────────────┤
│ Old Value:  nginx:1.⟦19⟧  (red strikethrough on "19")      │
│ New Value:  nginx:1.⟦21⟧  (green highlight on "21")        │
└─────────────────────────────────────────────────────────────┘
```

### PR DESCRIPTION
**Pull request by noamoss**
## Summary

This PR implements two visual UI improvements to enhance the diff viewing experience:

1. **Improved Section Movement Indication (#64)**: Enhanced visual indicators to clearly distinguish between path segment separators and movement transitions using double arrows (⇒) and explicit "from:" and "to:" labels.

2. **Character-Level Diff Highlighting (#74)**: Added character-level highlighting in Split View to show exactly which characters changed within a line, similar to GitHub's inline diffs.

## Related Issues

Fixes #64
Fixes #74

### Issue #64 - Section Movement Indication
- Added visual movement indicators with "from:" and "to:" labels
- Used double arrow (⇒) for movement transitions vs single arrow (→) for path segments
- Movement indicators only appear when cards are expanded (consistent with other change types)
- Applied consistent styling across Cards view and Split view sidebar
- Added ARIA labels for accessibility

### Issue #74 - Character-Level Diff Highlighting
- Integrated character-level diff computation using existing `diff-match-patch` library
- Added CodeMirror decorations for character-level highlighting:
  - Deleted characters: red background (#ffd7d9) with strikethrough
  - Added characters: green background (#ccfdf3) with underline
- Performance optimization: only computes diffs for changed lines
- Handles edge cases: empty strings, whitespace, special characters

### Testing Infrastructure
- Set up Playwright testing framework
- Created visual regression tests for movement indicators
- Created tests for character-level diff highlighting
- Added test utilities and fixtures

## Files Changed

- `ui/components/ChangeCard.tsx` - Enhanced movement indication
- `ui/components/SplitDiffView.tsx` - Added movement indicators and character-level diff
- `ui/components/InlineDiscussion.tsx` - Updated movement indicator format
- `ui/app/globals.css` - Added styles for movement and character diff indicators
- `ui/package.json` - Added Playwright dev dependency
- `ui/playwright.config.ts` - Playwright configuration
- `ui/tests/` - Test suite for visual regression and functionality

## Testing

- [x] TypeScript compilation passes
- [x] All existing functionality preserved
- [x] Visual testing completed manually
- [x] Playwright tests created (can be run with `npm test` in ui directory)

### Movement Indicators
- **Before**: Path arrows and movement arrows looked identical
- **After**: Clear distinction with double arrow (⇒) for movement, single arrow (→) for paths

### Character-Level Highlighting
- **Before**: Only line-level highlighting in Split View
- **After**: Character-level highlighting shows exact changes within lines

## Checklist

- [x] Tests added/updated
- [x] Code follows project style guidelines
- [x] Accessibility improvements (ARIA labels)
- [ ] Documentation updated (if needed)

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ui/components/ChangeCard.tsx [hunk 1]
```diff
@@ -342,11 +342,23 @@ export default function ChangeCard({ change, index, oldYaml, newYaml }: ChangeCa
...
+                className="flex items-center gap-2 text-xs text-gray-700 bg-purple-50 px-2 py-1 rounded border border-purple-200"
+                data-testid="movement-indicator"
+                role="status"
+                aria-label={`Section moved from ${formatMarkerPath(change.old_marker_path)} to ${formatMarkerPath(change.new_marker_path)}`}
+              >
+                <span className="text-purple-700 font-semibold">from:</span>
+                <span className="font-mono text-purple-800 bg-purple-100 px-1.5 py-0.5 rounded">
+                  {formatMarkerPath(change.old_marker_path)}
+                </span>
+                <span className="text-purple-600 font-bold text-sm" aria-hidden="true">⇒</span>
```
Comment (noamoss) [inline ui/components/ChangeCard.tsx]: **Accessibility: Unicode Arrow Character**

The double arrow (⇒) used on line 356 may not render correctly on all systems or screen readers. Consider using a more accessible alternative:

**Suggested Fix:**
1. Use an SVG arrow icon instead of Unicode
2. Or use CSS `::before`/`::after` pseudo-elements with proper ARIA labels
3. Ensure the arrow is properly announced by screen readers

The current `aria-hidden="true"` on the arrow is good, but the visual indicator should be more robust.

**Comment by noamoss** [comment]
## Branch Rebased - Clean Commit Created

I've rebased the branch onto `main` and created a clean commit with **only the UI improvements** for issues #64 and #74.

### What Changed:
- **Before**: PR included 124 files (rename + UI changes)
- **After**: PR now includes only 10 UI files (842 insertions, 43 deletions)

### Commit Structure:
- Rebased onto `main` (which already has the rename commit)
- Created new commit `44bb9f0` with only UI improvements:
  - Movement indicators (#64)
  - Character-level diff highlighting (#74)
  - Playwright test infrastructure

### Files Changed (UI only):
- `ui/components/ChangeCard.tsx`
- `ui/components/SplitDiffView.tsx`
- `ui/components/InlineDiscussion.tsx`
- `ui/app/globals.css`
- `ui/package.json` & `ui/package-lock.json`
- `ui/playwright.config.ts` (new)
- `ui/tests/` (new test files)

The PR is now focused solely on the UI improvements, making it much easier to review! 🎯

File: ui/components/ChangeCard.tsx (hunk 1 above)
Comment (noamoss) [inline ui/components/ChangeCard.tsx]: **Accessibility: Unicode Arrow Character**

The double arrow (⇒) on line 356 may not render correctly on all systems, fonts, or screen readers. While `aria-hidden="true"` is correctly applied, the visual indicator should be more robust.

**Suggested Improvement:**
Consider using an SVG arrow icon or CSS-based arrow for better cross-platform compatibility:

```tsx
<span className="text-purple-600 font-bold text-sm" aria-hidden="true">
  <svg width="16" height="16" viewBox="0 0 16 16" fill="currentColor">
    <path d="M8 0L16 8L8 16L6.5 14.5L12.5 8.5H0V7.5H12.5L6.5 1.5L8 0Z"/>
  </svg>
</span>
```

Or use CSS `::before`/`::after` pseudo-elements with a more reliable arrow character like `→` (single arrow) or `⟹` (double arrow with better font support).

File: ui/components/SplitDiffView.tsx [hunk 2]
```diff
@@ -228,7 +229,81 @@ export default function SplitDiffView({ oldYaml, newYaml, diff }: SplitDiffViewP
     return { oldHighlightLines: oldLines, newHighlightLines: newLines };
   }, [diff.changes]);
 
-  // Create decorations for old editor (using API semantic changes)
+  // Compute character-level diffs for changed lines
+  const [charDiffs, setCharDiffs] = useState<Map<string, DiffChunk[]>>(new Map());
+
+  useEffect(() => {
```
Comment (noamoss) [inline ui/components/SplitDiffView.tsx]: **Code Quality: Missing Cleanup for Async Operations**

The `useEffect` on lines 235-304 starts an async operation but doesn't handle cleanup if the component unmounts or dependencies change before the async operation completes. This could lead to:
1. Setting state on an unmounted component (React warning)
2. Race conditions if dependencies change quickly

**Suggested Fix:**
Add cleanup logic:

```typescript
useEffect(() => {
  let cancelled = false;
  
  const computeDiffs = async () => {
    // ... computation logic ...
    
    if (!cancelled) {
      setCharDiffs(diffs);
    }
  };

  computeDiffs();
  
  return () => {
    cancelled = true;
  };
}, [oldYaml, newYaml, diff.changes]);
```

**Review by noamoss — COMMENTED** [review]
## Code Review Summary (After Rebase)

Much better! The PR is now focused solely on UI improvements. The package rename has been removed, making this a clean feature PR. Here's my review:

### Issues Found

1. **Performance**: Character diff computation in `useEffect` could block UI for large documents. Consider batching or using `requestIdleCallback`.

2. **Testing Anti-pattern**: Tests use `waitForTimeout` instead of waiting for actual conditions. This makes tests flaky.

3. **CSS Consistency**: Character diff styles use hardcoded colors instead of CSS variables, inconsistent with movement indicator styles.

4. **Test Utility Bug**: `hasCharDiffHighlighting` uses incorrect selectors that don't match the actual CSS classes (`.cm-char-removed`, `.cm-char-added`).

5. **Async Cleanup**: Missing cleanup logic in `useEffect` for async operations could cause React warnings or race conditions.

6. **Offset Calculation**: Potential edge case where chunk lengths might not match line content exactly.

7. **Accessibility**: Unicode arrow (⇒) may not render correctly on all systems. Consider SVG or more reliable alternatives.

### Positive Aspects

✅ **Clean PR scope**: Focused only on UI improvements  
✅ **Good accessibility**: ARIA labels properly implemented  
✅ **Comprehensive testing**: Playwright tests cover both features  
✅ **Well-structured code**: Clear separation of concerns  
✅ **Type safety**: Good TypeScript usage throughout  
✅ **Visual improvements**: Movement indicators and character highlighting are well-implemented  
✅ **Performance consideration**: Only computes diffs for changed lines

### Recommendations

1. **Fix test selectors** in `hasCharDiffHighlighting` to match actual CSS classes
2. **Replace `waitForTimeout`** with proper Playwright waiting mechanisms
3. **Add CSS variables** for character diff colors for consistency
4. **Add cleanup logic** to the async `useEffect`
5. **Consider SVG arrow** instead of Unicode for better compatibility

The core implementation is solid. These are mostly polish and best-practice improvements that will make the code more maintainable and tests more reliable.

File: ui/components/SplitDiffView.tsx [hunk 3]
```diff
@@ -228,7 +229,93 @@ export default function SplitDiffView({ oldYaml, newYaml, diff }: SplitDiffViewP
...
+    let cancelled = false;
```
Comment (noamoss) [inline ui/components/SplitDiffView.tsx]: **Excellent improvement!** The async cleanup logic with the `cancelled` flag is now properly implemented. This prevents React warnings and race conditions. Great work addressing this concern!

File: ui/tests/helpers/test-utils.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+/**
+ * Find a change card by change type
+ */
+export async function findChangeCardByType(
+  page: Page,
+  changeType: string
+) {
+  return page.locator('[data-testid="change-card"]').filter({
+    hasText: new RegExp(changeType, 'i'),
+  }).first();
```
Comment (noamoss) [inline ui/tests/helpers/test-utils.ts]: **Perfect fix!** The `hasCharDiffHighlighting` function now uses the correct selectors (`.cm-char-removed, .cm-char-added`) that match the actual CSS classes used in the implementation. This will make the tests actually verify the correct elements.
